hls_vi/generate_indices.py

```diff
@@ -192,6 +199,31 @@ def apply_fmask(data: np.ndarray, fmask: np.ndarray) -> np.ma.masked_array:
     return np.ma.masked_array(data, fmask & cloud_like != 0)
 
 
+def apply_union_of_masks(bands: List[np.ma.masked_array]) -> List[np.ma.masked_array]:
+    """Mask all bands according to valid data across all bands
+
+    This is intended to reduce noise by masking spectral indices if
+    any reflectance band is outside of expected range of values ([1, 10000]).
+    For example the NBR index only looks at NIR and SWIR bands, but we might have
+    negative reflectance in visible bands that indicate the retrieval has issues
+    and should not be used.
+
+    Reference: https://github.com/NASA-IMPACT/hls-vi/issues/44
+    """
+    if not bands:
+        return []
+
+    # NB - numpy masked arrays "true" is a masked value, "false" is unmasked
+    # so bitwise "or" will  mask if "any" band has a masked value for that pixel
+    mask = bands[0].mask.copy()
+    for band in bands[1:]:
+        mask |= band.mask
+
+    for band in bands:
+        band.mask = mask
+    return bands
```

Comment:
Take it or leave it, as this is arguably less readable:

```suggestion
    from functools import reduce
    mask = reduce(np.ma.mask_or, bands, np.ma.nomask)

    for band in bands:
        band.mask = mask

    return bands
```

Alternatively, you can "spell out" what `reduce` does, but I would suggest at least avoiding the unnecessary copy and initialize the mask to nomask, and use mask_or for its potential performance bump (although, in this context, it may be unnoticeable):

```suggestion
    mask = np.ma.nomask

    for band in bands:
        mask = np.ma.mask_or(mask, band.mask)

    for band in bands:
        band.mask = mask

    return bands
```

Either way, this avoids the need for an explicit check on the length of `bands`.

Reply:
Thanks for the tip! Both are much nicer...

I went with the loop though because I found that the reduce variant needed to unpack the mask bands to get the dtypes correct,

```python
>>> mask = reduce(np.ma.mask_or, bands, np.ma.nomask)
...
ValueError: Incompatible dtypes 'bool'<>'int64'

>>> mask = reduce(np.ma.mask_or, [b.mask for b in bands], np.ma.nomask)
```

Thanks for the tip! I really like that this saves the check on "bands" length since the 0 length shouldn't actually happen so the check was only there for correctness. This is much nicer